scene/gui/button.cpp

```diff
@@ -72,287 +129,266 @@ void Button::_notification(int p_what) {
-				icon_ofs.x = 0.0;
-			}
-
-			int text_clip = size.width - style->get_minimum_size().width - icon_ofs.width;
-			if (_internal_margin[SIDE_LEFT] > 0) {
-				text_clip -= _internal_margin[SIDE_LEFT] + theme_cache.h_separation;
-			}
-			if (_internal_margin[SIDE_RIGHT] > 0) {
-				text_clip -= _internal_margin[SIDE_RIGHT] + theme_cache.h_separation;
-			}
-
-			text_buf->set_width(is_clipped ? text_clip : -1);
+			// Draw the text.
+			if (!xl_text.is_empty()) {
+				text_buf->set_alignment(align_rtl_checked);
 
-			int text_width = MAX(1, is_clipped ? MIN(text_clip, text_buf->get_size().x) : text_buf->get_size().x);
+				float text_buf_width = MAX(1.0f, drawable_size_remained.width); // The space's width filled by the text_buf.
+				text_buf->set_width(text_buf_width);
 
-			Point2 text_ofs = (size - style->get_minimum_size() - icon_ofs - text_buf->get_size() - Point2(_internal_margin[SIDE_RIGHT] - _internal_margin[SIDE_LEFT], 0)) / 2.0;
-
-			if (vertical_icon_alignment == VERTICAL_ALIGNMENT_TOP) {
-				text_ofs.y += icon_region.size.height / 2;
-			}
-			if (vertical_icon_alignment == VERTICAL_ALIGNMENT_BOTTOM) {
-				text_ofs.y -= icon_region.size.height / 2;
-			}
+				Point2 text_ofs;
 
-			text_buf->set_alignment(align_rtl_checked);
-			text_buf->set_width(text_width);
-			switch (align_rtl_checked) {
-				case HORIZONTAL_ALIGNMENT_FILL:
-				case HORIZONTAL_ALIGNMENT_LEFT: {
-					if (icon_align_rtl_checked != HORIZONTAL_ALIGNMENT_LEFT) {
-						icon_ofs.x = 0.0;
-					}
-					if (_internal_margin[SIDE_LEFT] > 0) {
-						text_ofs.x = style->get_margin(SIDE_LEFT) + icon_ofs.x + _internal_margin[SIDE_LEFT] + theme_cache.h_separation;
-					} else {
-						text_ofs.x = style->get_margin(SIDE_LEFT) + icon_ofs.x;
-					}
-					text_ofs.y += style->get_offset().y;
-				} break;
-				case HORIZONTAL_ALIGNMENT_CENTER: {
-					if (text_ofs.x < 0) {
-						text_ofs.x = 0;
-					}
-					if (icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_LEFT) {
-						text_ofs += icon_ofs;
-					}
-					text_ofs += style->get_offset();
-				} break;
-				case HORIZONTAL_ALIGNMENT_RIGHT: {
-					if (_internal_margin[SIDE_RIGHT] > 0) {
-						text_ofs.x = size.x - style->get_margin(SIDE_RIGHT) - text_width - _internal_margin[SIDE_RIGHT] - theme_cache.h_separation;
-					} else {
-						text_ofs.x = size.x - style->get_margin(SIDE_RIGHT) - text_width;
-					}
-					text_ofs.y += style->get_offset().y;
-					if (icon_align_rtl_checked == HORIZONTAL_ALIGNMENT_RIGHT) {
-						text_ofs.x -= icon_ofs.x;
+				switch (align_rtl_checked) {
+					case HORIZONTAL_ALIGNMENT_CENTER: {
+						text_ofs.x = (drawable_size_remained.width - text_buf_width) / 2.0f;
 					}
-				} break;
-			}
+						[[fallthrough]];
+					case HORIZONTAL_ALIGNMENT_FILL:
+					case HORIZONTAL_ALIGNMENT_LEFT:
+					case HORIZONTAL_ALIGNMENT_RIGHT: {
+						text_ofs.x += style_margin_left;
+						if (_internal_margin[SIDE_LEFT] > _internal_margin[SIDE_RIGHT]) {
```

Comment:
And then this makes the right margin to be ignored completely, even if it's set to be larger than the left one.

Reply:
If it exists on both sides, it can be calculated similar to stylebox.

Reply:
What I meant is that you've unified code for LEFT and RIGHT here, but this part only adjusts for the left margin. Compare this to how you handle icon's horizontal alignment.

Reply:
I set the width of `text_buf` to `drawable_size_remained.width`. This way the alignment can be left to `TextParagraph` and we can calculate the offset from the left.